### DISCUSSION TITLE
14 May Next Dev Session

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#1286] [General]
When: 14 May 2000 UTC+8/1300 CET/1200 GMT/0800 ET

Where: [via Matrix](https://matrix.to/#/#mesa-dev:matrix.org) (switching to Matrix Dev Session Channel)

Proposed Agenda:

- May milestones
- Namespace Decision
- Quartzsite Release (June)
- Complexity Explorer Tutorial
- NUMFOCUS Grant
- Deployment & PyPI
  - Deployment patterns? 
  - https://github.com/projectmesa/mesa/issues/1252

 @rht @wang-boyu @jackiekazil @Corvince @aiwhoo

**Comment by tpike3** [comment]
## May Dev Session Notes

On 14 May Mesa held is monthly Dev Session. Below are the notes, please correct me if I missed anything or failed to capture it properly. 

- **Namespace Decision**
     -  Based on user request and user feedback we voted to implement a flattening namespace, which allows users to import mesa and then directly leverage major classes. As an example, mesa.datacollector.DataCollector becomes mesa.DataCollector. It was also determined that for files like `time.py`, `visualization.py`, `space.py` context matters so context still exists as an example `mesa.space.MultiGrid` or `mesa.space.SingleGrid`  
     - Through the debate and due to low collision possibility we also agreed to implement a `mesa.flat` implementation as an example `mf.MultiGrid`, `mf.DataCollector`. We will revisit this `mesa.flat` in 6 months (November 2022) to reassess. 
     - @rht already implemented this via #1294 (Thanks @rht!)

 - **Quartzsite Release**
     - We are still on track for [Quartzsite release](https://github.com/projectmesa/mesa/milestone/25) next month. This will include a breaking change (#1300) where batch_run default number of processors is 1. This is to make it easier for new users and we are storing multi-processing `batch_run`  how-to as a [snippet](https://github.com/projectmesa/mesa/blob/main/docs/useful-snippets/snippets.rst).
 
- **Complexity Explorer**
    - Goal is to have examples of Session 2 - Introduction of SugarScape and [Session 3- Set up and initialize landscape](https://github.com/edgeofchaos42/ComplexityExplorer/blob/main/Session_3_Initial_Model_and_Add_Landscape.ipynb) to SFI by Quartzsite release. Introductory tutorial done by 1 September for fall term. 
 
- **NUMFOCUS Grant**
     - @tpike3 owes a starter draft to the dev team by 28 May. Broad agreement that grant should focus on #1263 and use a bounty or similar to attract someone to make Mesa-Jupyter interface.

- **Deployment and PyPI** 
    - @jackiekazil will research permissions for GH Actions and PYPI release to ensure Mesa resilience.  This will include looking at #1282 and #1076

**Comment by tpike3** [comment]
**Suspense: 4 JUNE**

So I learned I am not good at writing grant, a clear lack of experience so please feel free to rip this apart and make improvements.

Draft portions are in yellow. (seen in .pdf) 

@jackiekazil @rht 

[Mesa NumFOCUS Small Development Grant.txt](https://github.com/projectmesa/mesa/files/8791130/Mesa.NumFOCUS.Small.Development.Grant.txt)

[Mesa NumFOCUS Small Development Grant.pdf](https://github.com/projectmesa/mesa/files/8791131/Mesa.NumFOCUS.Small.Development.Grant.pdf)

**Reply by rht** [reply]
> Widgets so users can adjust model parameters from number of agents to distribution of attributes.

The user input widgets will be taken from Jupyter's builtin input widgets, so we don't need to develop those.

**Reply by rht** [reply]
Probably need to say that Mesa already has a standalone Tornado-based viz server, but it doesn't mesh well with Jupyter ecosystem.

> This grant will provide comparable tools in Python allowing users to leverage Python’s exponentially larger and more robust scientific ecosystem. As Mesa is Python organic it will also make the associated modelling ecosystem more extensible.

The non-GUI part of Mesa already can be used together with other scientific Python stack. The Jupyter GUI doesn't necessarily make it even more so. I'd say the main reason is that people (academic researchers? data scientists?) mostly do explorations in a Jupyter environment nowadays, and so we want to support this common use case.

**Reply by tpike3** [reply]
> > Widgets so users can adjust model parameters from number of agents to distribution of attributes.
> 
> The user input widgets will be taken from Jupyter's builtin input widgets, so we don't need to develop those.

Update to Proposal Description: 

*This grant will be used to integrate Mesa with Jupyter, so users can more easily create interactive visualizations of their models. This will include 2D and 3D canvas to show how the model is behaving and help convey the dynamics of the model for collaboration and presentation. Integration of Jupyter widgets with Mesa so users can adjust model parameters from number of agents to distribution of attributes. Charts so users can display critical statistics of the model to demonstrate behavior and validity of the model against real world phenomena.*

**Reply by tpike3** [reply]
> 

Update: 
The impact on the scientific ecosystem will be to improve ABMs as an emerging and powerful scientific tool. As Jupyter is a tool of choice for many academic researchers and data scientists, this will extend Mesa to more effectively leverage the Jupyter suite reducing user barriers to ABMs as a powerful computational approach to explore and understand their complex phenomena of interest. Furthermore, NetLogo due to its interface and ease of use dominates much of ABMs, but it is a cloistered ecosystem designed originally for teaching and is relatively small and difficult to connect. This grant will develop Mesa so it has comparable ease-of-use interface tools  so more people can leverage Python’s expansive scientific computing ecosystem. 


Full update: 
[Mesa NumFOCUS Small Development Grant (1).pdf](https://github.com/projectmesa/mesa/files/8793201/Mesa.NumFOCUS.Small.Development.Grant.1.pdf)

**Reply by jackiekazil** [reply]
@tpike3 is the last PDF the most recent version?

**Reply by tpike3** [reply]
Yes

**Reply by tpike3** [reply]
Last day for comments, it is due tomorrow.

**Reply by jackiekazil** [reply]
done.

**Reply by tpike3** [reply]
Grant submitted (I took out emails but attached is the version we submitted)
[Mesa NumFOCUS Small Development Grant (2).pdf](https://github.com/projectmesa/mesa/files/8840349/Mesa.NumFOCUS.Small.Development.Grant.2.pdf
)
